### ISSUE TITLE
Hosting Upgrades

### ISSUE THREAD

**Issue by larsdesigns** [backdrop-ops/backdropcms.org#479]
Currently the virtual machines are using CentOS 6 with PHP 5.x. I propose performing the following  upgrades. 

- [x] Build new virtual machines in the same hosting account using Debian 9. Some benefits of Debian is that we can upgrade the OS via command line in the future, the Debian community values are more inline with that of  the Backdrop CMS community, and we can take advantage of the more modern repositories that are available. 
- [x] Install and configure PHP 7.2, MariaDB 10, NGINX, and the rest of the stack. 
- [x] Migrate the existing web applications to the new hosts. 
- [ ] Improve the scheduled backups to include off-site storage such as AWS S3 or something similar.
- [ ] Implement Let's Encrypt SSL certificates.
- [ ] Build an automated backup system for the application filesystem, and database dumps. Backup to an external source such as Amazon S3 or something similar.

**Comment by serundeputy** [comment]
I'm a `debian` fan.

**Comment by serundeputy** [comment]
I am for all of this.

@quicksketch might have thoughts on CentOS vs. Debian. Personally I prefer Debian.

**Comment by klonos** [comment]
I also prefer Debian/Ubuntu, but out of personal habit. No matter what, we should update to php7, and I would actually like to also see ZenCI testing against php 7.2.

**Comment by larsdesigns** [comment]
I am planning to have these servers built next month and go live before January. I will need to team up with the web folks to assist with the transfer of the sites. Just to make sure we get everything over.

**Comment by larsdesigns** [comment]
One of my clients suffered a data loss due to a Linode hardware failure. After three days of downtime and an expectation for the client to perform low level disk recovery tasks that proved fruitless, they are leaving Linode with frustration. 

Linode shows a list of daily backups in the backup page of their control panel. This mislead the client to believe that the backups listed would be available if needed. They are not. Only three backups are available via Linode, the last backup, a backup from ten days before, and a backup from three days before. All the other backups are not available even by request. The backup that was needed before the disk failed was not available in this case.

I have a long history with Linode and have not had problems, but this incident is making me reconsider the choice of hosting provider or at least reinforcing a need to have off-site backups like scheduled database and filesystem dumps to Amazon S3/Glacier or something similar for Backdrop CMS.

**Comment by larsdesigns** [comment]
I added bullet number 6 to the issue description calling for creating an external backup system.

**Comment by quicksketch** [comment]
I'm completely OK with switching off CentOS. I prefer Ubuntu these days, but Debian is completely okay with me. Note that Debian 9 (Stretch) includes PHP 7.0 by default, so we'll need to use a different upstream.

Also one more item, I'd love if we could start managing our servers with Puppet or similar, so we can track the software that is actually installed on these machines. As it is now, it's difficult to tell exactly how much work we have to do to move over all of our sites.

**Comment by quicksketch** [comment]
Also for the time being, I would love getting through items 1-3 ASAP and complete other items as we can.

**Comment by larsdesigns** [comment]
@quicksketch, Perfect timing. I have time to work on this over the weekend. I was planning on installing PHP 7.2.

Concerning Puppet, I have an installation script (playbook) that I use via Ansible. But I think I would prefer using Docker and share the images, Docker build files, etc with the community. This would make it self documenting, transparent, and potentially help onboard developers to contribute. However, I have found using reverse proxy configuration to be tricky in Docker.

I am happy to build the environment in Docker. But I am afraid of over engineering. 

Many years ago, I have used Puppet and I can dust off those skills if need be.

**Comment by serundeputy** [comment]
I'd vote for in order of preference:
  1. Docker
  2. Ansible
  3. Puppet

**Comment by quicksketch** [comment]
That'd be awesome @larsdesigns! You already have a Linode user account. Feel free to create a new server as the current size. We haven't had any scaling problems that I'm aware of.

My preference is to not use docker for a production environment. Ansible is fine with me. Can we commit the playbook scripts here to this repository so we can keep track of what's happened on the server?

**Comment by larsdesigns** [comment]
1. I created the new Linode host in NJ location and I am working on the playbooks. I chose New Jersey because the other Linodes are in New Jersey. But if we are to use California or a different location this is the time to make the change.
2. Signed up the automated daily image backups.

**Comment by larsdesigns** [comment]
Completed the following tasks: 

1. Installed PHP 7.2 FPM and common found PHP packages. 
1. Installed and configured IP Tables with rules for ipv4 and ipv6. 
1. Installed fail2ban but not configured
1. Installed MariaDB, performed the secure installation, and set root password. 
1. Installed nginx and it is running at  http://66.175.208.83 but not yet configured for PHP.
1. Copied the repo, sanitized_databases, and backups from CentOS server.
1. Installed fish
1. Installed emacs
1. Installed Ansible
1. Installed persistent tables and rebooted to see firewall rule load on boot. 
1. Created users for Jen and Nate and installed their public ssh keys. 
1. Installed net-tools. 
1. Installed git. 
1. Installed NRPE server. 
1. Installed Certbot and certbot nginx.
1. Installed postfix.

TO DO: 
1. Migrate the databases. 
1. Create database users for the sites.
1. Configure PHP
1. Configure Nginx
1. Optimize MariaDB configuration
1. Configure reverse proxy cache for Nginx. 

I have kept a log of what was installed specifically in an Ansible playbook. I will post it when I am done.

**Comment by larsdesigns** [comment]
1. Completed the migration of the sites to the new VM. 
1. Re-used as much nginx configuration as possible. 
1. Checked the nginx cache on the sites and all the sites are caching except for the dcu.backdropcms.org but I found in the configuration file that the cache is set to expire immediately on purpose. This could probably be improved.
1. I migrated the entire "/home" directory from the centos host. 
1. Created all of the user accounts that I found and restored authorized_keys and id_rsa files. 
1. Migrated the existing SSL certificates and they are working fine in Firefox without DNS. 
1. To help with testing, make the following entry in your '/etc/hosts' file: 
```
66.175.208.83 new.backdropcms.org www.backdropcms.org backdropcms.org forum.backdropcms.org, api.backdropcms.org, dcu.backdropcms.org.
```
1. Formatted all of the nginx sites-available files so they are easier to read using two spaces instead of tabs, cleaning up line endings, etc. 
1. The new.backdropcms.org url returns the `phpinfo()` and is temporary until we launch. 
1. Because mcrypt was removed from PHP 7.2 and moved to PECL because of licensing issues, I had install it using PECL. It should be okay for a while be consider using a different encryption method in the future: http://php.net/manual/en/intro.mcrypt.php. 

TO DO: 
1. [  ] Configure NRPE. 
1. [  ] Configure automated offsite backups. 
1. [  ] Coordinate with @quicksketch for cutover.

**Comment by larsdesigns** [comment]
@quicksketch, I did install memcached but I did not configure it. I found your memcached documentation: https://gist.github.com/quicksketch/3070c757482f2335d72917223957bc76. Would you like me to configure it and created the bins?

**Comment by larsdesigns** [comment]
@quicksketch and @jenlampton, I did not set the hostname yet. What would you like it to be?

**Comment by larsdesigns** [comment]
1. [ ] Decide on a hostname. 
1. [ ] Decide on Memcached implementation.  
1. [ ] Decide when to launch. 
1. [ ] Migrate cron jobs. Oops I forgot...

**Comment by larsdesigns** [comment]
> 1. Completed the migration of the sites to the new VM.
> 2. Re-used as much nginx configuration as possible.
> 3. Checked the nginx cache on the sites and all the sites are caching except for the dcu.backdropcms.org but I found in the configuration file that the cache is set to expire immediately on purpose. This could probably be improved.
> 4. I migrated the entire "/home" directory from the centos host.
> 5. Created all of the user accounts that I found and restored authorized_keys and id_rsa files.
> 6. Migrated the existing SSL certificates and they are working fine in Firefox without DNS.
> 7. To help with testing, make the following entry in your '/etc/hosts' file:
> 
> ```
> 66.175.208.83 new.backdropcms.org www.backdropcms.org backdropcms.org forum.backdropcms.org, api.backdropcms.org, dcu.backdropcms.org.
> ```
> 1. Formatted all of the nginx sites-available files so they are easier to read using two spaces instead of tabs, cleaning up line endings, etc.
> 2. The new.backdropcms.org url returns the `phpinfo()` and is temporary until we launch.
> 3. Because mcrypt was removed from PHP 7.2 and moved to PECL because of licensing issues, I had install it using PECL. It should be okay for a while be consider using a different encryption method in the future: http://php.net/manual/en/intro.mcrypt.php.
> 
> TO DO:
> 
> 1. [ ]  Configure NRPE.
> 2. [ ]  Configure automated offsite backups.
> 3. [ ]  Coordinate with @quicksketch for cutover.
> 4. [ ]  Provide Ansible Playbook and supporting documentation.

@jenlampton Can you check and see if the sites are running okay on the new VM?

**Comment by jenlampton** [comment]
> Decide on a hostname.

Is there any reason we can't use backdropcms.org for the hostname? If not, let's go with that.

**Comment by jenlampton** [comment]
We don't need dcu.backdropcms.org anymore, we can probably take that site down entirely. But we do need localize.backdropcms.org, which I don't see in your list yet for testing.

**Comment by larsdesigns** [comment]
@jenlampton, I enabled the localize.backdropcms.org site. 

/etc/hosts

```
66.175.208.83 new.backdropcms.org backdropcms www.backdropcms.org backdropcms.org forum.backdropcms.org api.backdropcms.org dcu.backdropcms.org localize.backdropcms.org
```

For the hostname, how about www2.backdropcms.org? The host that we are replacing has the name www1.backdropcms.org.

I removed the nginx entry for dcu.backdropcms.org. Would you like me to remove the database and filesystem also?

**Comment by jenlampton** [comment]
www2 works! let's leave the DB and files until @klonos has a chance to
object :)

**Comment by larsdesigns** [comment]
@jenlampton, I set the hostname and renamed the VM within the Linode account.

**Comment by quicksketch** [comment]
After today's meeting, we arranged to switch the API site first. With @larsdesigns's help, we've now cut over to the Debian / PHP 7.2 server!

![image](https://user-images.githubusercontent.com/100206/55207768-ce53fb80-5198-11e9-9022-744779173e51.png)

We also cut over the utility server sanitize.backdropcms.org (where we hold sanitized copies of the content database). For now this means that if you need a dump of the API site, you use sanitize.backdropcms.org. If you need any of the other sites, you use sanitize-old.backdropcms.org.

**Comment by larsdesigns** [comment]
@quicksketch, When would you like to perform the next cut over?

**Comment by larsdesigns** [comment]
The sanitize databases and files are not updating. I am debugging it now.

**Comment by larsdesigns** [comment]
I made some adjustments to the file permissions and recreated the symlinks. It is running again: https://sanitize.backdropcms.org/sanitized/tmp_db2-April-1-2019-84829-sanatized.sql.gz.

I will check back late to make sure that the cron job actually runs on its own.

**Comment by klonos** [comment]
> let's leave the DB and files until @klonos has a chance to object :)

I object!! 😅

I meant to be using this site, but then it was suggested that we should implement some GitHub webhooks thingy instead ...which never happened.

Another thing that would help would be to allow me to be able to merge and deploy for dcu if that is not a problem. I would then be able to be more "caring" with it.

As an alternative, I could spin this up in a personal linode, so i could be managing it (would still need to retain merge access to the dcu GitHub repo). In that case, all I would need would be some DNS changes to point dcu.b.org to the IP of that linode.

LMK

**Comment by jenlampton** [comment]
@klonos I don't think anyone's done anything with it since we "launched", so I'd rather not maintain it as a b.org project. If we send you the db and files (and IIRC you already have access to the codebase?) then you can put it wherever you like. Mayb having a local copy would be enough?

**Comment by klonos** [comment]
@jenlampton yep, that would do I guess. I'll probably set it up somewhere public instead of locally though. Reasons being that a) I would like cronjobs to kick in and send notification emails b) would like others to help maintain it, and be added to the list of people that receive those notification emails.

That means that I will have to incur some cost, but I guess that a cheap $5/mo linode would do.

**Comment by jenlampton** [comment]
I don't want you to incur any cost, I was just thinking that since we are not using this it doesn't make sense to maintain it. Why don't we see if we can start using it first, and if we actually do, then we set it back up on the b.org server?

**Comment by larsdesigns** [comment]
I am available tonight and tomorrow for cut-overs. 

@jenlampton @quicksketch

**Comment by jenlampton** [comment]
I can be available tonight @larsdesigns. I'll hop on Gitter.

**Comment by jenlampton** [comment]
@larsdesigns I'm seeing 504 Gateway Time-out errors on backdropcms.org.

**Comment by larsdesigns** [comment]
@jenlampton I think backdropcms.org is running on the old host.

**Comment by larsdesigns** [comment]
Thanks for the sprint last night. I am so happy that the forum.backdropcms.org and localize.backdropcms.org have been moved to the new host. 

My next steps are to setting up NRPE and post the Ansible playbook for the build.

**Comment by quicksketch** [comment]
We're hoping to do the last domain (backdropcms.org) tonight. Similarly, we'll get everything all set on the new server, put the old site into maintenance mode, dump the database again, and import into the new server. BackdropCMS.org sees more traffic than the other sites, so we'll want to ensure we don't allow anyone to create content during the down window.

**Comment by larsdesigns** [comment]
@quicksketch and I finished migrating the backdropcms.org site to the new VM.

**Comment by larsdesigns** [comment]
To Do: 

1. [ ] Setup Nagios monitoring. 
1. [ ] Configure offsite backups.

**Comment by larsdesigns** [comment]
Finished implementing Nagios and NRPE. I will see if I can work out off-site backups this week.

**Comment by larsdesigns** [comment]
I have attached the Ansible playbook that I used to record what was installed on the new host and how it was installed.

[backdropcms_playbook.yml.txt](https://github.com/backdrop-ops/backdropcms.org/files/3726579/backdropcms_playbook.yml.txt)

**Comment by larsdesigns** [comment]
The virtual machine www1-backdropcms-org has been offline for a while. With approval from the stakeholders, may I proceed in deleting the virtual machine at Linode. 

@jenlampton

**Comment by jenlampton** [comment]
yes, go ahead and delete it.

**Comment by klonos** [comment]
🔥

**Comment by larsdesigns** [comment]
I have deleted it.

**Comment by larsdesigns** [comment]
The Linode VM backups seem to be adequate for now. The Lindoe SSL certificates have been implemented in #220. 

No further tasks. Closing...
